src/SourceLocation.js

```diff
@@ -0,0 +1,42 @@
+// @flow
+import type {LexerInterface} from "./Token";
+
+/**
+ * Lexing or parsing positional information for error reporting.
+ * This object is immutable.
+ */
+export default class SourceLocation {
+    lexer: LexerInterface; // Lexer holding the input string.
+    start: number;         // Start offset, zero-based inclusive.
+    end: number;           // End offset, zero-based exclusive.
+
+    constructor(lexer: LexerInterface, start: number, end: number) {
+        this.lexer = lexer;
+        this.start = start;
+        this.end = end;
+        Object.freeze(this); // Immutable to allow sharing in range().
```

Comment:
Apparently this impacted performance in the past, but modern browsers seem to be unaffected.

Reply:
Ok, LMK if you'd like me to remove.

Reply:
Let's leave it for now.